design/design.md

```diff
@@ -0,0 +1,380 @@
+```
+Versioning for gRPC is usually done by changing the package name, which creates
+protobuf objects in a new c++ namespace. For example, changing the package name
+to `xrp.v2` results in objects with namespace `xrp::v2`. In the event
+that a handler does not change between versions (which implies the objects are
+the same in all but namespace), the old handler can be reused by templating the
+handler signature like so:
+
+```
+template<class T, class R>
+R doFee(RPC::ContextGeneric<T>& context);
+```
+
+## Interface
+The interface to the GRPC service is defined by the .proto files. The `.proto`
+files define the structure of the messages, the methods available, their inputs
+and outputs, and the service itself.  GRPC, with the help of protobuf,
+autogenerates code based on these `.proto` files, for both client and server.
+GRPC has bindings in many common programming languages
+([link](https://grpc.io/docs/reference/)).
+A client of the GRPC service we are building would use the exact same `.proto`
+files as the server, to generate the client code. The generated code (both
+client and server), handles serialization, deserialization and I/O. Protobuf
+allows one to define messages with named and typed fields, and allows for user
+defined types. Client and server code can then manipulate the generated objects using the
+message names, field names and types defined in the `.proto` file. For example,
+using the below example `.proto` file:
+```
+service XRPLedgerAPI {
+  // Get account info for an account on the XRP Ledger.
+  rpc GetAccountInfo (GetAccountInfoRequest) returns (AccountInfo);
+
+  // Get the fee for a transaction on the XRP Ledger.
+  rpc GetFee (GetFeeRequest) returns (Fee);
+  message GetAccountInfoRequest {
+    // The address to get info about.
+    bytes address = 1;
+  }
+
+  message AccountInfo {
+    XRPAmount balance = 1;
+  
+    uint32 sequence = 2;
+  
+  }
+  message XRPAmount {
+    // A numeric string representing the number of drops of XRP.
+    bytes drops = 1;
+  }
+}
+```
+A client can do things like:
+```
+GetAccountInfoRequest request;
+request.set_address("rLkMJhSVwhmummLjJPVrwQRZZYiYQhVQ1A");
+//Send the request, receive populated response
+uint32_t sequence = response.sequence();
+std::string drops = response.balance.drops();
+```
+Note, the `bytes` type in a protobuf message is decoded as `std::string` in C++.
+
+For more information about protocol buffers see
+[here](https://developers.google.com/protocol-buffers/docs/overview).
+For more information about gRPC and protocol buffers, see
+[here](https://grpc.io/docs/guides/)
+
+To support gRPC inside rippled, existing handlers will be mapped 1:1 to `.proto`
+files and protobuf handlers; for each RPC method, there will exist a protobuf message
+with the same fields and data that the existing handlers are expecting in JSON form,
+as well as a handler that accepts this protobuf message.
```

Comment:
Leaving a note that I want to think about this approach. Some of our json interfaces are overly complicated, and I'm not sure we want to replicate those interfaces exactly.

Reply:
That makes sense; this might be somewhat of a philosophical question, but how would you go about determining which part of the interface to replicate and what to not replicate?

Reply:
@cjcobb23 A couple things we could do is 1) Talk to the team (especially Elliot, Rome, and Nik) to see if people have existing lists of pain points in the existing API they'd like to resolve; 2) read the API docs and keep an eye out for things that are hard to explain 3) Look for parameters or commands that try to do too many things (the multiple ways of encoding secrets, for example) 

I do understand there are huge benefits to just using the existing API, and I expect that we will end up doing that. However, this is an opportunity to try to fix some of the pain points in the current API. We should at least consider it.